aten/src/ATen/Context.h

```diff
@@ -390,6 +393,7 @@ class TORCH_API Context {
   bool enabled_mathSDP = true;
   bool enabled_cudnnSDP = true;
   bool enabled_overrideable = true;
+  bool allow_fp16_bf16_reduction_mathSDP = false;
```

Comment:
Cc @jbschlosser 
I think the default value should be switched. Since many people have voiced concerns over performance we should revert to the old behavior and make people turn manually set this to true

Reply:
Agree with Driss - imo let's revert to the old behavior as the default to fix the perf regression and make the high-precision behavior opt-in.

Reply:
After quite a bit of internal discussion, I'm adjusting my stance here. To summarize:
* We have precedent for providing high precision for composite ops by default and making low precision opt-in. The original PR can be considered a "bug fix" to bring consistency in following this precedent.
* I'm reading `allow_fp16_bf16_reduction_mathSDP` as "opt-in to low precision". So I think the default value should be `false`.

Unfortunately, this does mean those expecting the old behavior + performance have to explicitly change this flag.

Reply:
>     I'm reading `allow_fp16_bf16_reduction_mathSDP` as "opt-in to low precision". So I think the default value should be `false`, in-line with the "bug fix" from the original PR.

So because the `allow_fp16_bf16_reduction_mathSDP` flag name implies "opt-in to low precision" it should be high precision by default?
The flag that was introduced in this same PR?
The flag we could rename right now before it lands in `release/2.5.0` and causes issue reports titled "*Help I updated the app and lost half my speed 6800 XT*" to open up on downstream repositories?

Reply:
Why not something easy to understand like `torch.set_sdpa_math_precision` keeping in spirit with `torch.set_float32_matmul_precision`? Then it could be "medium" or whatever you want by default.

Reply:
@eqy Good point: we can have separate configurations on FP16/BF16 matmul reduced precision off/on with torch.backends.cuda.matmul.allow_fp16_reduced_precision_reduction = False/True. Currently the nested options are available for numerical testing.

For the original intention of https://github.com/pytorch/pytorch/pull/128922 , it's for improving the accuracy (and "Math SDPA reduced precision off" brings the best accuracy.

Reply:
@Beinsezii  We can use other potential options (with better namings). Feel free to suggest better flag names.

> We have precedent for providing high precision for composite ops by default and making low precision opt-in.

From https://github.com/pytorch/pytorch/issues/76440 , https://github.com/pytorch/pytorch/pull/76509 , it matches what @jbschlosser  mentioned, we used  "opt-in to low precision" as a tradition in PyTorch (e.g., use TF32 for fp32 matmul, or use fp16/bf16 for math backend in SDPA), while the default one is the high precision (e.g., FP32 for fp32 matmul, and fp32 for math backend in SDPA).

Reply:
@jianyuh If my understanding is correct, the `scaled_dot_product_attention` function prefers the `FLASH_ATTENTION` and `EFFICIENT_ATTENTION` over the `MATH` kernels by default, despite the former two backends using bf16/fp16 like the old `MATH`. So if you're opting into pure `MATH` manually by disabling the fast kernels, to me it makes sense to also opt-into high precision ops. What having fp32 upcast MATH as the default effectively does is make the naive behavior of `scaled_dot_product_attention` use fast ops on NVIDIA and the MI300, but high precision ops on all the other cards which is an anti-pattern.

Reply:
@Beinsezii One issue for using fp16/bf16 for math backend (platform independent) is that it has much worse accuracy compared to flash attention or efficient attention backend.

Check https://github.com/pytorch/pytorch/pull/128922#issuecomment-2201176008

https://github.com/pytorch/pytorch/pull/128922#issuecomment-2251960439

Using high precision MATH makes the accuracy match more on flash attention or efficient attention with FP16/BF16.

In this PR, we intend to provide the *option* for the user to switch back to the less accurate but more efficient math backend, as you suggested.